File: python-version/LC869/reordered_power_of_2.py
class Solution:
    def __init__(self):
        self.__visited = None

    def reorderedPowerOf2(self, n: int) -> bool:
        chars = str(n)
        self.__visited = [False] * len(chars)
        return self.__backtrack(chars, 0, 0)

    def __backtrack(self, chars: str, idx: int, num: int) -> bool:
        if idx == len(chars):
            return self.__powerOf2(num)
        for i in range(0, len(chars)):
            if chars[i] == '0' and num == 0:
                continue
            if self.__visited[i]:
                continue
            self.__visited[i] = True
            if self.__backtrack(chars, idx+1, num * 10 + ord(chars[i]) - ord('0')):
                return True
            self.__visited[i] = False
        return False

    @staticmethod
    def __powerOf2(n: int) -> bool:
        return n > 0 and (n & (n - 1)) == 0

File: go-version/leetcode/LC869/reordered_power_of_2.go
package LC869

import "strconv"

/**
 * @Author: cutejiuge cutejiuge@163.com
 * @Date: 2025/8/10 下午10:52
 * @FilePath: reordered_power_of_2
 * @Description: 回溯
 */

var visited []bool

func reorderedPowerOf2(n int) bool {
	chars := []byte(strconv.Itoa(n))
	visited = make([]bool, len(chars))
	return backtrack(chars, 0, 0)
}

func backtrack(chars []byte, idx, num int) bool {
	if idx == len(chars) {
		return powOf2(num)
	}
	for i := 0; i < len(chars); i++ {
		if num == 0 && chars[i] == '0' {
			continue
		}
		if visited[i] {
			continue
		}
		visited[i] = true
		if backtrack(chars, idx+1, num*10+int(chars[i]-'0')) {
			return true
		}
		visited[i] = false
	}
	return false
}

func powOf2(n int) bool {
	return n > 0 && (n&(n-1)) == 0
}
